radicle-crdt/src/gmap.rs

```diff
@@ -9,7 +9,7 @@ use crate::Semilattice;
 /// Conflicting elements are merged via the [`Semilattice`] instance.
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct GMap<K, V> {
-    inner: BTreeMap<K, V>,
+    pub inner: BTreeMap<K, V>,
```

Comment:
I'm not sure about the `pub` for this because it breaks encapsulation.

Reply:
Yeah, I was expecting a comment like that ;)
IMO encapsulation is an inadequate relic of the shared mutable state era of OOP languages.  If you have the type system protecting your from violating invariants (by the way if mut/non-mut refs), encapsulation is just a nuisance

Reply:
> Yeah, I was expecting a comment like that ;)

Hehehe

> IMO encapsulation is an inadequate relic of the shared mutable state era of OOP languages. If you have the type system protecting your from violating invariants (by way of mut/non-mut refs), encapsulation is just a nuisance

In this case, it's not about any shared mutable state. It's about disallowing the user of the type from accessing the internal data structure due to the CRDT semantics. A `GMap` is a "Grow-only Map" emphasising the _Grow_. If the library user suddenly has access to the `BTreeMap` then they are free to use its delete functionality. Now our CRDT semantics are out the window :disappointed:

Reply:
Right, a valid point.  I’ll remove those ‘pub’ declarations 👍